### PR TITLE
Move SMJ join filtered part out of join_output stage. LeftOuter, LeftSemi

### PR DESCRIPTION
**Pull request by comphead**
## Which issue does this PR close?

Related to #11555 
Related to #12359  

Closes #.

## Rationale for this change

Move filtered logic out of join_output stage.

The problem is for filtered joins there is extra step need, e.g to calculate final filtered mask for the specific row the algorithm, it needs the knowledge for  every right row processed for the given left row. 

For example for LeftOuter join :
`select * from t1 join t2 on (t1.a = t2.a and t1.b > t2.b)`

t1:
|a|b|
|---|---|
|1| 10|

t2:
|a|b|
|---|---|
|1| 9|
|1|11|

Currently the Join will output

|a|b|a|b
|---|---|---|---|
|1| 10|||
|1| 10|1|11|

which is incorrect, the first null matched row shouldn't be out as for given row there is a match exists.

The problem is records calculated depending on batch size and the output can be called anytime once the output size hit the batch size. So with `batch size == 1` in this scenario the first row will be out because output == 1 which is equal to batch_size but the algorithm on this stage have no idea that another matched row is coming. 

The idea of algorithm is to move filtered algorithm out of join stage and be not dependent on batch size. Instead it will be called once left row index switched to next index, to be sure every right row is processed for given left row, thus it will be still in batches although not strictly equal to batch_size.

### CODE REVIEW

**Comment by comphead** [comment]
@korowa @viirya Please have a look on prototype with just Left Outer join moved out of join output stage. Let me know your thoughts. With this approach I hope we can handle all other filtered cases like right, semi, anti

**Comment by comphead** [comment]
@alamb cc

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:709]: we need more information to track the filtered bitmask across incoming batches

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:745]: NULL is the row shouldnot go to the output

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:747]: false means for null joined rows

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:873]: this block is magic, but without it the join stucks, I believe it is something with output sizes. We can improve it later

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:985]: we need to track batch ids, to prevent edge case like
Batch 0 value 1 streamed index 0
Batch 1 value 2 streamed index 0

So the streamed index is the same, and without batch id its not possible to figure out the actual values are different

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1436]: clones will be removed

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1497]: For Left WIP we pass the original non filtered batch because filtering will be done later.

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1503]: Collect all the neccessary extra information across batches

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1518]: Its not needed it doesn't have left join, hopefully we can remove it

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1643]: for filtered joins we dont need to clear batches as we need them later, I'll do it rhough the flag later

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1648]: here is the main logic: But in fact what it does it just consolidate batches, filtering information and does the filtering in very similar way we currently do in `freeze_streamed`

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs
Comment (comphead) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:142]: to be removed

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs
Comment (comphead) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:137]: this test passes now

**Comment by korowa** [comment]
@comphead, I also was checking for these two issues, and found that currently SMJ contains some utility logic used in other join operators, but reimplements it in its own way. In addition the data processing could be implemented in more similar to other joins fashion -- currently it's kind of hard to track what's going on there.

I'm currently trying to rework it [here](https://github.com/korowa/arrow-datafusion/tree/smj-rework), and I've got fuzz tests passing (but there are other issues), but it still needs some cleanups, proper working with spills and comments/docs.

**Comment by comphead** [comment]
> @comphead, I also was checking for these two issues, and found that currently SMJ contains some utility logic used in other join operators, but reimplements it in its own way. In addition the data processing could be implemented in more similar to other joins fashion -- currently it's kind of hard to track what's going on there.
> 
> I'm currently trying to rework it [here](https://github.com/korowa/arrow-datafusion/tree/smj-rework), and I've got fuzz tests passing (but there are other issues), but it still needs some cleanups, proper working with spills and comments/docs.

Thanks @korowa the way how SMJ implemented is truly hard to understand, it took me a couple of painful weeks to dig through it.

I feel propose to go into 2 directions? 
To unblock some projects that already relies on DF SMJ , we can fix the current  SMJ Leftouter and other filtered variants with approach above or simialer and in parallel have your new implementation reviewed and onboarded?

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:789]: filtering phase happens when the left row index gets changed to ensure all right rows processed for the given row

**Comment by korowa** [comment]
> I feel you propose to go into 2 directions?

Sure, we can, they are not mutually exclusive and, likely, current approach will be completed and delivered faster, as it changes less things. I just wanted to say that (ideally, if possible) this fix could potentially be achieved by already reusing utility functions from other join implementations (like `adjust_indices_by_join_type`).

Regarding this fix -- it seems now there are two places doing almost the same thing (please correct me if I'm wrong)
1) filtering batches and correcting filter mask while staging output batches (applying filter and `get_filtered_join_mask` in `freeze_all`)
2) filtering batches and correcting mask while producing output batches (`filter_joined_batch` which calls `get_corrected_filter_mask`)

+ `get_filtered_join_mask` is, likely, intended to do exactly the same thing as `get_corrected_filter_mask` due to its comment line:
> /// This return a tuple of:
/// - corrected mask with respect to the join type

So, maybe there is a chance, that current implementation of filtering and aligning output indices according to the join type logic can be located in a single place for all join types, using the same functions, rather than being spreaded across two processing stages?

**Comment by comphead** [comment]
Thanks @korowa for the feedback, this is not a final PR, the goal was to discuss the direction, having LeftOuter filtered join as an example. 

If we okay with the example I'm planning to move all the filtered logic into new single place and make the code cleaner. 

Thanks for `adjust_indices_by_join_type` hint, I'll check if we can reuse it after getting the code restructured like mentioned above, most likely in separate PR to make things manageable.

**Comment by comphead** [comment]
@korowa I'm planning to move all other join variants to the same approach so the filtered logic will be in a single place, test it out and make the PR ready for your review

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs
Comment (comphead) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:230]: will be enabled when Left Anti has moved out of join partial as well

**Comment by comphead** [comment]
@korowa @viirya @alamb can I get the review please on this on the final PR moving the filtering SMJ logic out of join partial phase. Its done for Left/LeftSemi/Inner Join.

Right Joins, Full, Anti I'm planning to move right after this PR has approved

**Comment by alamb** [comment]
I will plan to review this either later today or tomorrow

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:913]: there is multiple if statements of this kind just because we moved only 2 types of join

**Comment by comphead** [comment]
> I will plan to review this either later today or tomorrow

Thanks @alamb

File: datafusion/sqllogictest/test_files/sort_merge_join.slt
Comment (alamb) [inline datafusion/sqllogictest/test_files/sort_merge_join.slt:103]: does this mean that if we merge this PR the answers are incorrect (aka that we will be introducing a regression for some period of time?)

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (alamb) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:706]: perhaps we can comment this field too

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (alamb) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:709]: Could you please comment what the fields represent -- In particular how is `filter_mask`, `row_indices` and `batch_ids`  interpreted relative to the `batches` (do they always have the same row count? What does the batch represent?

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (alamb) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:794]: I think it would be good to list out the possibilities here -- I think it is only `JoinType::Inner` right?

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (alamb) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:1013]: Can you please put that context as comments of the docs of `stream_batch_counter`?

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (alamb) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:1463]: do you still plan to remove the clones?

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (alamb) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:828]: I wonder if adding some methods on `JoinedRecordBatches` would make this code easier to read - For example instead of directly accessing the fields, perhaps adding functions would allow 

```suggestion
                                        if !self.output_record_batches.is_empty()
```

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (alamb) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:898]: Maybe you could change it to `continue` to avoid outputting an empty batch 🤔

**Review by alamb — COMMENTED** [review]
Thank you @comphead -- I think this seems like a nice improcement to me. I left some stylistic comments that I think would help encapsulation, etc

My only real concern about this PR is the commented out tests. I noticed that you say

> Please do not put much attention on disabled tests, as before the filtered join was unstable anyway

I don't understand what this means. Do you mean that while the tests passed, the actual output of SortMergeJoin was likely not correct and thus the code with this change will not be any less correct. 

I think once that is sorted out this PR would be good to go

File: datafusion/sqllogictest/test_files/sort_merge_join.slt
Comment (comphead) [reply datafusion/sqllogictest/test_files/sort_merge_join.slt:103]: SMJ is experimental is disabled by default. Even now depending on data distribution every variant of filtered SMJ join is not stable and gives incorrect result, despite the fact that  simple tests passed, that's the reason the fuzz tests are commented out for filtered variants.

I understand the concern, but moving all variants to new approach will make the PR unmanageable to review. I'll try to move other variants as fast as possible, moreover the most popular variants(Inner, Left, LeftSemi) are moved in this PR

**Comment by comphead** [comment]
Thanks @alamb for the feedback I'll address them asap

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:794]: correct, perhaps Full Join won't also need a separate mask as Full Join is a specific case of Left and Right.

File: datafusion/physical-plan/src/joins/sort_merge_join.rs
Comment (comphead) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:828]: That totally makes sense, I feel it requires more methods to add(add/is_empty/clear), I'd prefer to do it as follow up, this PR is too large imho

File: datafusion/sqllogictest/test_files/sort_merge_join.slt
Comment (comphead) [reply datafusion/sqllogictest/test_files/sort_merge_join.slt:103]: The issues https://github.com/apache/datafusion/issues/11555
https://github.com/apache/datafusion/issues/12359 focus on SMJ correctness issues. Fuzz tests flaky because of SMJ has a problem do answering your question this PR is to make SMJ to work correctly

**Review by alamb — APPROVED** [review]
This seems like an improvement to me - thank you @comphead 

Let's get this in and keep iterating

**Pull request merged**
